processing/src/main/java/io/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java

```diff
@@ -0,0 +1,394 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.groupby.epinephelinae;
+
+import com.google.common.collect.Ordering;
+import com.google.common.primitives.Ints;
+import io.druid.java.util.common.ISE;
+
+import java.nio.ByteBuffer;
+import java.util.Comparator;
+
+/**
+ * ByteBuffer-based implementation of the min-max heap developed by Atkinson, et al.
+ * (http://portal.acm.org/citation.cfm?id=6621), with some utility functions from
+ * Guava's MinMaxPriorityQueue.
+ */
+public class ByteBufferMinMaxOffsetHeap
```

Comment:
Hm, I could see that, I'll leave this specialized for the current use case in this PR though, if another use comes up later we can do a PR then